File: 프로그래머스/3/43238. 입국심사/입국심사.cs
using System;

public class Solution {
    public long solution(int n, int[] times) {
        long left = times[0];
        long right = times[times.Length-1]*(long)n;
        
        while(left<=right){
            long mid = (left+right)/2;
            
            long cnt = 0;
            for(int i=0; i<times.Length; i++){
                cnt += (mid/times[i]);
                if(cnt >= n) break;
            }
            
            if(cnt >= n) right = mid-1;
            else left = mid +1;
        }
        return left;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

long long solution(int n, vector<int> times) {
    //sort(times.begin(), times.end());

    long long mid, left = times[0], right = (long long)times.back() * n;

    while (left <= right) {
        mid = left + (right - left) / 2;

        long long possibleClientCnt = 0;
        for (int i = 0; i < times.size(); i++) {
            possibleClientCnt += (mid / times[i]);
            if (possibleClientCnt >= n) break;
        }
        if (possibleClientCnt < n) {  // 현재 기준 시간으로는 모든 고객을 받지 못하므로 기준 시간을 높여야함
            left = mid + 1; // left 갱신
        }
        else {
            right = mid - 1;
        }
    }
    return left;
}